### PR TITLE
Binary installer workflow

### PR DESCRIPTION
**Pull request by mauwii**
# DO NOT MERGE YET

### CODE REVIEW

File: .github/workflows/binary-installers.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  pip-compile:
+    strategy:
+      fail-fast: false
+      matrix:
+        output-file:
+          - py3.10-darwin-x86_64-reqs
+          - py3.10-linux-x86_64-cuda-reqs
+          - py3.10-windows-x86_64-cuda-reqs
+        python-version:
+          - '3.9'
```
Comment (tildebyte) [inline .github/workflows/binary-installers.yml:20]: Why?

File: .github/workflows/binary-installers.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: debug output
+        run: |
+          pip3 -V
+          python3 -V
+          pip3 list
+
+      - name: run pip-compile
+        id: pip-compile
+        run: pip-compile --allow-unsafe --generate-hashes --output-file=binary_installer/${{ matrix.output-file }}.txt binary_installer/requirements.in
```
Comment (tildebyte) [inline .github/workflows/binary-installers.yml:57]: This is not something which should be run frequently. Dependent package versions can change *daily*. We should only be recompiling from scratch if

1) a dependent package *must* be updated (it's Corrupting data, there's a security issue, etc.)
2) (possibly) something *we* do makes us incompatible with a specific package version and we need to move to a compatible one
3) we *add* a dependent package (and even in that case this command is incorrect: the correct command includes `reuse-hashes`)

*Removing* a dependent package likewise only requires deleting the dependency from requirements.in and the various requirements files (no pip-compile is required at all)

Either way, the full command for generating ALL dependencies is `pip-compile -r --annotate --annotation-style split --generate-hashes --emit-index-url --allow-unsafe --resolver=backtracking --output-file=binary_installer/${{ matrix.output-file }}.txt binary_installer/requirements.in`; the full command for UPDATING is `pip-compile -r --annotate --annotation-style split --generate-hashes --emit-index-url --allow-unsafe --resolver=backtracking --reuse-hashes --output-file=binary_installer/${{ matrix.output-file }}.txt binary_installer/requirements.in` (note that the only change is *adding* `--reuse-hashes`)

File: .github/workflows/binary-installers.yml (hunk 1 above)
Comment (mauwii) [reply .github/workflows/binary-installers.yml:20]: As already written to you in Discord, this is for Debug Reason (maybe you also saw the big fat "DO NOT MERGE" in the first comment ...)

**Comment by mauwii** [comment]
BTW: It would be hard to bring this workflow into a more useable form while windows is failing 😅

File: .github/workflows/binary-installers.yml (hunk 2 above)
Comment (mauwii) [reply .github/workflows/binary-installers.yml]: @tildebyte so should I use the `generate all deps` or the `update deps` command? If I understand you right, point 3 mentions to use the `update deps` version, right?


Problem for Windows generation of reequirements currently is

```
There are incompatible versions in the resolved dependencies:
  basicsr==1.4.1 (from gfpgan==1.3.8->-r binary_installer/requirements.in (line 29))
  basicsr>=1.4.2 (from realesrgan==0.3.0->-r binary_installer/requirements.in (line 15))
```

which of course is also not fixed by neither one or the other of the commands you mentioned (I just tested this in my fork now).

Comment (tildebyte) [reply .github/workflows/binary-installers.yml]: > BTW: My plan is to run this command only when changes to the requirements.in file have been done (which would be the only reason to update the requirement files)

That's perfect

**Comment by tildebyte** [comment]
> what's the reason for the pinned GFPGAN Verison?

IDK, You'd have to ask @lstein

**Comment by lstein** [comment]
> > what's the reason for the pinned GFPGAN Verison?
> 
> IDK, You'd have to ask @lstein

The pinned GFPGAN version is the wrong workaround for a problem some windows users encountered. The symptom was that `basicsr` 1.4.2 failed to install. The incorrect solution was to pin a customized version of GFPGAN in which its setup had been doctored to import `basicsr` 1.4.1. The correct solution (discovered today!) is to enable Long Paths in the Windows registry. Apparently 1.4.2 has a file with a long pathname, while 1.4.1 doesn't.

I've unpinned GFPGAN in the latest iteration of the unified directory installer. That can be done here too.

**Comment by lstein** [comment]
@mauwii I should probably put the binary installer into a usable state such that it installs into the `invokeai` directory rather than into the repo. Are there other issues that need to be addressed before this PR is merged?

**Comment by mauwii** [comment]
@lstein: Yes, there is a lot of discussion open, for example:

- when should this workflow run (obviously not on every PR)
- when should it include `--reuse-hashes`
- where to puth the created zip files:
    - for the moment they're attached to the workrun run itself
    - should they push to the repo
    - should they add the attachments to the release
    - ...

So the DO NOT MERGE YET headline is serious imho, so will convert this PR to draft to make it more obvoius

**Comment by mauwii** [comment]
I guess this one is not relevant anymore

**Pull request closed without merging**